src/index.ts

```diff
@@ -113,30 +126,44 @@ const extension: JupyterFrontEndPlugin<void> = {
       setting.changed.connect(updateSettings);
     }
 
-    let cellStartTime = new Date();
-
     NotebookActions.executionScheduled.connect((_, args) => {
+      const { cell, notebook } = args;
       if (enabled) {
-        cellStartTime = new Date();
+        cellExecutionMetadataTable[cell.model.id] = {
+          index: notebook.activeCellIndex,
+          scheduledTime: new Date()
+        };
       }
     });
 
     NotebookActions.executed.connect((_, args) => {
       if (enabled) {
-        const { cell, notebook, success, error } = args;
         const cellEndTime = new Date();
+        const { cell, notebook, success, error } = args;
+        const cellId = cell.model.id;
+        const notebookId = notebook.id;
+        const scheduledTime = cellExecutionMetadataTable[cellId].scheduledTime;
+        const recentExecutedCellTime =
+          recentNotebookExecutionTimes[notebookId] || scheduledTime;
+        cellExecutionMetadataTable[cellId].startTime =
+          scheduledTime >= recentExecutedCellTime
+            ? scheduledTime
+            : recentExecutedCellTime;
+        cellExecutionMetadataTable[cellId].endTime = cellEndTime;
+        recentNotebookExecutionTimes[notebookId] = cellEndTime;
```

Comment:
it may be worth using an in-memory lru cache like https://www.npmjs.com/package/lru-cache to limit the number of notebook execution times.

Reply:
decided to use lru-cache for `recentNotebookExecutionTime`.